packages/react-slider/src/components/Slider/useSliderState.tsx

```diff
@@ -226,41 +260,37 @@ export const useSliderState = (state: SliderState) => {
 
   const originPercent = origin ? getPercent(origin, min, max) : 0;
 
-  const thumbStyles = {
-    transform: vertical ? `translateY(${valuePercent}%)` : `translateX(${valuePercent}%)`,
+  const thumbWrapperStyles = {
+    transform: vertical
+      ? `translateY(${valuePercent}%)`
+      : `translateX(${dir === 'rtl' ? -valuePercent : valuePercent}%)`,
     transition: stepAnimation ? `transform ease-in-out ${animationTime}` : 'none',
+    ...state.thumbWrapper.style,
+  };
+
+  const thumbStyles = {
+    transform: dir === 'rtl' ? 'translate(50%, -50%)' : 'translate(-50%, -50%)',
     ...state.thumb.style,
   };
 
-  const trackStyles = vertical
-    ? {
-        top: origin ? `${Math.min(valuePercent, originPercent)}%` : 0,
-        height: origin
-          ? `${Math.max(originPercent - valuePercent, valuePercent - originPercent)}%`
-          : `${valuePercent}%`,
-        borderRadius:
-          origin && origin !== (max || min)
-            ? `${originPercent > valuePercent ? '99px 99px 0px 0px' : '0px 0px 99px 99px'}`
-            : '99px',
-        transition: stepAnimation
-          ? `transform ease-in-out ${animationTime}, height ease-in-out ${animationTime}`
-          : 'none',
-        ...state.track.style,
-      }
-    : {
-        left: origin ? `${Math.min(valuePercent, originPercent)}%` : 0,
-        width: origin ? `${Math.max(originPercent - valuePercent, valuePercent - originPercent)}%` : `${valuePercent}%`,
-        borderRadius:
-          origin && origin !== (max || min)
-            ? `${originPercent > valuePercent ? '99px 0px 0px 99px' : '0px 99px 99px 0px'}`
-            : '99px',
-        transition: stepAnimation
-          ? `transform ease-in-out ${animationTime}, width ease-in-out ${animationTime} ${
-              origin ? ', left ease-in-out ' + animationTime : ''
-            }`
-          : 'none',
-        ...state.track.style,
-      };
+  const trackStyles = {
+    [vertical ? 'top' : dir === 'rtl' ? 'right' : 'left']: origin ? `${Math.min(valuePercent, originPercent)}%` : 0,
+    [vertical ? 'height' : 'width']: origin
+      ? `${Math.max(originPercent - valuePercent, valuePercent - originPercent)}%`
+      : `${valuePercent}%`,
+    borderRadius:
+      origin && origin !== (max || min)
+        ? vertical
+          ? `${originPercent > valuePercent ? '99px 99px 0px 0px' : '0px 0px 99px 99px'}`
+          : (dir === 'rtl' ? valuePercent > originPercent : originPercent > valuePercent)
+          ? '99px 0px 0px 99px'
+          : '0px 99px 99px 0px'
+        : '99px',
+    transition: stepAnimation
+      ? `transform ease-in-out ${animationTime}, ${vertical ? 'height' : 'width'} ease-in-out ${animationTime}`
+      : 'none',
```

Comment:
These computations look pretty complex... Did you consider to extract them into a separate function and covering it with unit tests?

Reply:
That is a good idea, I'll look into separating and testing them.